src/pause/138CC0.c

```diff
@@ -103,19 +108,393 @@ s32 pause_badges_scroll_offset_x(s32 x) {
     return x;
 }
 
-INCLUDE_ASM(s32, "pause/138CC0", pause_badges_try_remove);
+s32 pause_badges_try_remove(s16 badgeID) {
+    s16 *currentSlot = gPlayerData.equippedBadges;
+    s16 *slotToRemove = currentSlot;
+    s32 result = 0;
+    s32 i;
+
+    if (badgeID == BADGE_INVALID) {
+        return 0;
+    }
+
+    for (i = 0; i < ARRAY_COUNT(gPlayerData.equippedBadges); i++, currentSlot++) {
+        if (badgeID == *currentSlot) {
+            slotToRemove = currentSlot;
+            result = 1;
+            break;
+        }
+    }
 
-INCLUDE_ASM(s32, "pause/138CC0", pause_badges_try_equip);
+    // Remove and shift array contents
+    for (; i < ARRAY_COUNT(gPlayerData.equippedBadges)-1; i++, currentSlot++) {
+        *currentSlot = *(currentSlot+1);
+    }
+    *currentSlot = 0;
 
-INCLUDE_ASM(s32, "pause/138CC0", pause_badges_draw_bp_orbs);
+    if (*slotToRemove == 0) {
+        result = 2;
+    }
+    return result;
+}
+
+typedef s32 BadgeEquipResult;
+enum BadgeEquipResults {
+    EQUIP_RESULT_ALREADY_EQUIPPED,
+    EQUIP_RESULT_NOT_ENOUGH_BP,
+    EQUIP_RESULT_TOO_MANY_BADGES,
+    EQUIP_RESULT_SUCCESS,
+};
```

Comment:
Should be able to do this:
```suggestion
typedef enum BadgeEquipResult {
    EQUIP_RESULT_ALREADY_EQUIPPED,
    EQUIP_RESULT_NOT_ENOUGH_BP,
    EQUIP_RESULT_TOO_MANY_BADGES,
    EQUIP_RESULT_SUCCESS,
} BadgeEquipResult;
```
Good enum btw :+1:

Reply:
That actually causes a mismatch unless I introduce a cast to `s32` in the switch in `pause_badges_handle_input`. It looks like the optimizer notices that there's only 4 valid enum values and gets rid of checks for <0 and >3 in the switch conditions.

Maybe it'd be better for `pause_badges_try_equip` to return s32 and still drop the typedef